### PR TITLE
Fix #252

### LINKED ISSUE #252: Improve the dependency resolver and its errors
**Is your feature request related to a problem? Please describe.**
Check out [this commit](https://github.com/valory-xyz/open-autonomy/pull/1160/commits/eb6fb33273be770101a06e2de72738219e457569). It has a circular dependency in `ledger` connection. The error thrown when running `make generators` at this commit is not helpful

**Describe the solution you'd like**
Throw a useful error that there's a circular dependency caused by the package in question.

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

Fix #252 by providing a useful error message when a circular dependency is detected while resolving the dependency tree.

Copy-pasting commit message from 357430a as it provides a good summary:

> This commit includes the following improvements to the DependencyTree.resolve_tree method,
while trying not to detour too much from the old structure:
> - split the method in two methods: one is the entry point to the recursive procedure
  (DependencyTree.resolve_tree), and the other is an auxiliary method used by the
  entry point which implements the actual recursive procedure.
> - The entry point, from the client code perspective, is now a pure function:
  given an adjacency list of the dependency graph, return the same graph
  in the form of a nested dictionary.
> - modified the recursive method by doing only one recursive step per method call,
  rather than two steps as it was before this commit. This change gives better
  code structure and readability.
> - in case of a circular dependency, detect the cycle and print a useful error message.
  In case the cycle is actually a self-loop, specify it in the error message;
  Otherwise, print the entire chain that witnesses the circular dependency.
> - Improved docstrings and type hinting.

## Fixes

Fix #252

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [X] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [X] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [X] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [X] Lint and unit tests pass locally with my changes and CI passes too
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [X] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

n/a

### CODE REVIEW

File: aea/helpers/dependency_tree.py [hunk 1]
```diff
@@ -96,7 +97,7 @@ def to_package_id(public_id: str, package_type: str) -> PackageId:
     return PackageId(package_type, PublicId.from_str(public_id)).without_hash()
 
 
-class DependecyTree:
+class DependencyTree:
```
Comment (marcofavorito) [inline aea/helpers/dependency_tree.py:100]: just a renaming to fix a minor typo

File: aea/helpers/dependency_tree.py [hunk 2]
```diff
@@ -117,18 +118,61 @@ def get_all_dependencies(item_config: Dict) -> List[PackageId]:
...
+            visited.add(root_package)
+            stack.append(root_package)
+
+            # build the root of the subtree according to the dependency list
             root_dependencies: List[PackageId] = dependency_list.get(root_package, [])
-            for package in root_dependencies:
-                tree[root_package][package] = {
-                    p: {} for p in cast(list, dependency_list.get(package))
-                }
-                cls.resolve_tree(dependency_list, tree[root_package][package])
```
Comment (marcofavorito) [inline aea/helpers/dependency_tree.py:131]: this for-loop implemented a second recursive step in the same call of the recursive function. Removed as I thought could possibly improve readability and code structure (in particular, note that if remained as before I would have had to append-to/pop-from the stack before/after the recursive call in the inner for-loop, which looks fishy). Unless I back the wrong horse, the behaviour of the method should be equivalent.

**Comment by marcofavorito** [comment]
@angrybayblade @DavidMinarsch the PR can be reviewed

File: aea/helpers/dependency_tree.py (hunk 1 above)
Comment (marcofavorito) [reply aea/helpers/dependency_tree.py:100]: :warning: it is a breaking change

**Review by DavidMinarsch — APPROVED** [review]

**Review by angrybayblade — APPROVED** [review]
LGTM

**Pull request merged**